src/ray/object_manager/connection_pool.h

```diff
@@ -0,0 +1,148 @@
+  ///
+  /// \param object_directory The object directory, used for obtaining
+  /// remote object manager connection information.
+  /// \param connection_service The io_service the connection pool should use
+  /// to create new connections.
+  /// \param client_id The ClientID of this node.
+  ConnectionPool(ObjectDirectoryInterface *object_directory,
+                 asio::io_service *connection_service, const ClientID &client_id);
+
+  /// Register a receiver connection.
+  ///
+  /// \param type The type of connection.
+  /// \param client_id The ClientID of the remote object manager.
+  /// \param conn The actual connection.
+  void RegisterReceiver(ConnectionType type, const ClientID &client_id,
+                        std::shared_ptr<ReceiverConnection> &conn);
+
+  /// Remove a receiver connection.
+  ///
+  /// \param type The type of connection.
+  /// \param client_id The ClientID of the remote object manager.
+  /// \param conn The actual connection.
+  void RemoveReceiver(ConnectionType type, const ClientID &client_id,
+                      std::shared_ptr<ReceiverConnection> &conn);
+
+  /// Get a sender connection from the connection pool.
+  /// The connection must be released or removed when the operation for which the
+  /// connection was obtained is completed.
+  ///
+  /// \param type The type of connection.
+  /// \param client_id The ClientID of the remote object manager.
+  /// \param success_callback The callback invoked when a sender is available.
+  /// \param failure_callback The callback invoked if this method fails.
+  /// \return Status of invoking this method.
+  ray::Status GetSender(ConnectionType type, ClientID client_id,
+                        SuccessCallback success_callback,
+                        FailureCallback failure_callback);
+
+  /// Releasex a sender connection, allowing it to be used by another operation.
+  ///
+  /// \param type The type of connection.
+  /// \param conn The actual connection.
+  /// \return Status of invoking this method.
+  ray::Status ReleaseSender(ConnectionType type, SenderConnection::pointer conn);
+
+  /// Remove a sender connection. This is invoked if the connection is no longer
+  /// usable.
+  /// \param type The type of connection.
+  /// \param conn The actual connection.
+  /// \return Status of invoking this method.
+  // TODO(hme): Implement with error handling.
+  ray::Status RemoveSender(ConnectionType type, SenderConnection::pointer conn);
+
+ private:
+  /// A container type that maps ClientID to a connection type.
+  using SenderMapType =
+      std::unordered_map<ray::ClientID, std::vector<SenderConnection::pointer>,
+                         ray::UniqueIDHasher>;
+  using ReceiverMapType =
+      std::unordered_map<ray::ClientID, std::vector<std::shared_ptr<ReceiverConnection>>,
+                         ray::UniqueIDHasher>;
+
+  /// Adds a receiver for ClientID to the given map.
+  void Add(ReceiverMapType &conn_map, const ClientID &client_id,
```

Comment:
Looking briefly at the code, I don't think we even need these methods, but in the future, please try to define these as helper methods in an anonymous namespace in the .cc file instead of defining them as private methods.

Same for the `Remove` and `Count` methods below.

Reply:
I'm not sure I understand why these methods should live in an anonymous namespace in the `.cc` file vs. being private methods. They are in fact private methods of this class that access private members.